### PR TITLE
Add __all__ to some packages

### PR DESCRIPTION
**Pull request by ghost**
This PR adds `__all__` to the `__init__.py` files to avoid some unwanted imports when using `from sympy import *`.

See also #13525.

### CODE REVIEW

**Comment by asmeurer** [comment]
Some test failures. I would run pyflakes on each `__init__.py` file to make sure they are consistent.

**Comment by ylemkimon** [comment]
Restarted a build, since it seems to be a Travis error. Please fix merge conflicts. Thank you.

**Comment by asmeurer** [comment]
This would be easier to read if they were organized like this (for example, with `assumptions/__init__.py`):

```py
from .assume import AppliedPredicate, Predicate, AssumptionsContext, assuming
__all__ = ['AppliedPredicate', 'Predicate', 'AssumptionsContext', 'assuming']

from .ask import Q, ask, register_handler, remove_handler
__all__ += ['Q', 'ask', 'register_handler', 'remove_handler']

from .refine import refine
__all__ += ['refine']
```

This style is still recognized by pyflakes.

File: sympy/core/__init__.py [hunk 1]
```diff
@@ -1,34 +1,93 @@
...
+__all__ += [
+    "Lambda", "WildFunction", "Derivative", "diff", "FunctionClass",
+    "Function", "Subs", "expand", "PoleError", "count_ops",
+    "expand_mul", "expand_log", "expand_func",
+    "expand_trig", "expand_complex", "expand_multinomial", "nfloat",
+    "expand_power_base", "expand_power_exp"
+]
+
 from .evalf import PrecisionExhausted, N
+__all__ += ["PrecisionExhausted", "N", "evalf"]
```
Comment (pavel-tkachenko) [inline sympy/core/__init__.py]: `evalf` is still imported because it used in the main `__init__.py` by `evalf._create_evalf_table()`.

Comment (asmeurer) [reply sympy/core/__init__.py]: I would import the function in the main `__init__.py` file and then delete it after it is called.

**Comment by pavel-tkachenko** [comment]
ping @asmeurer @ylemkimon

File: sympy/__init__.py [hunk 2]
```diff
@@ -85,7 +85,8 @@ def __sympy_debug():
 from .printing import python, print_python, srepr, sstr, sstrrepr
 from .interactive import init_session, init_printing
 
-evalf._create_evalf_table()
+from .core.evalf import _create_evalf_table
+_create_evalf_table()
```
Comment (asmeurer) [inline sympy/__init__.py]: You need to `del` it here so it doesn't get imported.

File: sympy/matrices/__init__.py [hunk 3]
```diff
@@ -3,25 +3,58 @@
...
+]
 
-from .expressions import (MatrixSlice, BlockDiagMatrix, BlockMatrix,
-        FunctionMatrix, Identity, Inverse, MatAdd, MatMul, MatPow, MatrixExpr,
-        MatrixSymbol, Trace, Transpose, ZeroMatrix, blockcut, block_collapse,
-        matrix_symbols, Adjoint, hadamard_product, HadamardProduct,
-        Determinant, det, DiagonalMatrix, DiagonalOf, trace, DotProduct)
+MutableDenseMatrix = MutableMatrix = Matrix
+SparseMatrix = MutableSparseMatrix
+__all__ += ["MutableDenseMatrix", "SparseMatrix"]
```
Comment (asmeurer) [inline sympy/matrices/__init__.py]: This should also include MutableMatrix.

**Review by asmeurer — COMMENTED** [review]
Before merging we should double check that no imported names were accidentally removed. Running `pyflakes` on all the `__init__.py` files should be sufficient.

File: sympy/__init__.py (hunk 2 above)
Comment (pavel-tkachenko) [reply sympy/__init__.py]: I think it is not necessary because `_create_evalf_table()` is started with underscore.

File: sympy/__init__.py [hunk 4]
```diff
@@ -11,27 +11,24 @@
 
 """
 
+__all__ = []
```
Comment (gxyd) [inline sympy/__init__.py]: this is raising an error here
```python
Traceback (most recent call last):
  File "./bin/test", line 105, in <module>
    import sympy
  File "/home/gxyd/dev/sympy/sympy/__init__.py", line 16
    from __future__ import absolute_import, print_function
SyntaxError: from __future__ imports must occur at the beginning of the file
```

**Comment by gxyd** [comment]
Currently:
```python
Traceback (most recent call last):
  File "./bin/test", line 105, in <module>
    import sympy
  File "/home/gxyd/dev/sympy/sympy/__init__.py", line 62, in <module>
    from . import polys
  File "/home/gxyd/dev/sympy/sympy/polys/__init__.py", line 5, in <module>
    from . import polytools
  File "/home/gxyd/dev/sympy/sympy/polys/polytools.py", line 19, in <module>
    from sympy.polys.polyclasses import DMP
  File "/home/gxyd/dev/sympy/sympy/polys/polyclasses.py", line 10, in <module>
    from sympy import oo
ImportError: cannot import name oo
```

This potentially indicates a circular import.

**Comment by pavtkach** [comment]
ping @gxyd

Some weird failures.

**Comment by Abdullahjavednesar** [comment]
@asmeurer can you please have a look, Thanks.

**Comment by sidhantnagpal** [comment]
It seems the fork has been deleted. For anyone wishing to complete this, executing
`git fetch origin pull/13600/head:pr-13600` will create a new branch `pr-13600` (assuming origin points to sympy/sympy). The conflicts will need resolution as well.

**Comment by asmeurer** [comment]
I would first see if @0xPavelTkachenko is OK with using these changes. It may be best to just redo this from the start if you want to fix this issue.

**Pull request closed without merging**